parse/src/main/java/com/parse/ParsePlugins.java

```diff
@@ -169,27 +176,170 @@ InstallationId installationId() {
+        ParseFileUtils.getAllNestedFiles(
+            getOldParseDir().getAbsolutePath(),
+            filesToBeMigrated
+        );
+        runParseDirFilesMigrationToCacheDir(filesToBeMigrated);
+    }
+
+    public static String replaceLast(String text, String regex, String replacement) {
+        return text.replaceFirst("(?s)" + regex + "(?!.*?" + regex + ")", replacement);
+    }
+
+    /**
+     * Since there is two new different APIs {@link #getFilesDir()}, {@link #getCacheDir()} for caching.
+     * <p></p>
+     * - Config related files, that manages by SDK itself should be migrated to {@link #getFilesDir()}`.
+     * <p></p>
+     * - Others files, that exists in the {@link #getOldParseDir()} should be migrated to {@link #getCacheDir()}.
+     * <p></p>
+     * To identify the config files from {@link #getOldParseDir()},
+     * we are using a hard coded list which defines the SDK config files names.
+     *
+     * @param allFiles Files that needed to be migrated into new location.
+     */
+    private void runParseDirFilesMigrationToCacheDir(ArrayList<File> allFiles) {
+        if (allFiles.isEmpty()) {
+            return;
+        }
+        boolean useFilesDir = false;
+        //Hard coded config file names list.
+        ArrayList<String> configFilesToMigrateToFilesDir = new ArrayList<>();
+        configFilesToMigrateToFilesDir.add("installationId");
+        configFilesToMigrateToFilesDir.add("currentUser");
+        configFilesToMigrateToFilesDir.add("currentConfig");
+        configFilesToMigrateToFilesDir.add("currentInstallation");
+        configFilesToMigrateToFilesDir.add("LocalId");
+        configFilesToMigrateToFilesDir.add("pushState");
+        //Start migration for each files in `allFiles`.
+        for (File itemToMove : allFiles) {
+            try {
+                for (String configName : configFilesToMigrateToFilesDir) {
+                    if (itemToMove.getAbsolutePath().contains(configName)) {
+                        useFilesDir = true;
+                        break;
+                    } else {
+                        useFilesDir = false;
+                    }
+                }
+                //Parse the old sub directory name where the file should be moved (new location) by following the old sub directory name.
+                String oldSubDir = replaceLast(itemToMove.getAbsolutePath().replace(getOldParseDir().getAbsolutePath(), "").replace("/" + itemToMove.getName(), ""), itemToMove.getName(), "");
+                File fileToSave = new File(
+                    (useFilesDir ? applicationContext.getFilesDir() : applicationContext.getCacheDir())
+                        + "/com.parse/" +
+                        oldSubDir,
+                    itemToMove.getName());
+                //Perform copy operation if file doesn't exist in the new directory.
+                if (!fileToSave.exists()) {
+                    ParseFileUtils.copyFile(itemToMove, fileToSave);
```

Comment:
Can we add log somewhere here what was migrated? I believe it will be interesting in a verbose mode for the developers.

Reply:
Yes, Added (`PLog.v()`) verbos log for events like _Copying_, _Already Exist_, _Deletion_, _Migration_Completion_.